File: quickSort.c
#include <stdio.h>

void QuickSort(int *data, int left, int right)
{
	int temp = data[left];
	int p = left;
	int i = left, j = right;

	while(i <= j)
	{
		while(data[j] >= temp && j >= p)
			j--;
		if(j >= p){
			data[p] = data[j];
			p = j;
		}

		while(data[i] <= temp && i <= p)
			i++;
		if(i <= p)
		{
			data[p] = data[i];
			p = i;
		}
	}
	data[p] = temp;
	if(p-left > 1)
		QuickSort(data, left, p-1);
	if(right-p > 1)
		QuickSort(data, p+1, right);
}

int main()
{
	int intArray[5] = {10, 1, 100, 78, -1};
	QuickSort(intArray, 0, 4);
	int index;
	for(index = 0; index < 5; index++)
		printf("%d	", intArray[index]);

	printf("\n");

	return 0;
}

File: quickSort.py
#!/usr/bin/env python

def quickSort(theSeq):
	n = len(theSeq)
	recQuickSort(theSeq, 0, n-1)

def recQuickSort(theSeq, first, last):

	if first >= last:
		return
	else:
		pos = partitionSeq(theSeq, first, last)
		recQuickSort(theSeq, first, pos - 1)
		recQuickSort(theSeq, pos + 1, last)

def partitionSeq(theSeq, first, last):
	pivot = theSeq[first]
	left = first + 1
	right = last
	while left < right:
		while left < right and theSeq[left] < pivot:
			left += 1
		while right >= left and theSeq[right] >= pivot:
			right -= 1
		if left < right:
			theSeq[left], theSeq[right] = theSeq[right], theSeq[left]

	if right != first:
		theSeq[first] = theSeq[right]
		theSeq[right] = pivot
	return right

def shortQuickSort(theSeq):
	if not theSeq :
		return []
	return shortQuickSort([item for item in theSeq[1:] if item < theSeq[0]]) + theSeq[0:1] + \
		shortQuickSort([item for item in theSeq[1:] if item >= theSeq[0]])

if __name__ == '__main__':
	testList = [43, 12, 4124, 423, 48,421]
	print testList
	quickSort(testList)
	for item in testList:
		print item,
	print
	testList = [432, 312, 43, 412, 98, 90]
	print testList
	testList = shortQuickSort(testList)
	for item in testList:
		print item,
	print
